dog_classifier/net/train.py

```diff
@@ -40,40 +60,58 @@ def trainNN(training_parameters):
 
+    df_train = pd.read_csv(path_to_labels + 'train_labels.csv')
+    df_val = pd.read_csv(path_to_labels + 'val_labels.csv')
+    with K.tf.device('/cpu:0'):
+        trainDataloader = DataGenerator(df_train, encoder_model,
+                                        batch_size=bs_size)
+        valDataloader = DataGenerator(df_val, encoder_model,
+                                      batch_size=bs_size)
+
     model.compile(loss='categorical_crossentropy', optimizer='adam',
                   metrics=['accuracy'])
 
     earlystopper = EarlyStopping(monitor='val_loss',
                                  patience=early_stopping_patience,
                                  min_delta=early_stopping_delta,
                                  verbose=1)
+
     reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.2,
-                                  patience=3, verbose=1, min_lr=1e-6)
+                                  patience=3, verbose=1, min_lr=1e-7)
+    # History callback function seperate the model history and the model itself
+    # This useful because we can plot/save the history of a model after a
+    # KeyboardInterrupt
+    hist = History()
 
-    history = model.fit_generator(trainDataloader, validation_data=valDataloader,
-                                  epochs=num_of_epochs,
-                                  callbacks=[earlystopper, reduce_lr])
+    modelCheckpoint = ModelCheckpoint(filepath=model_save_path + '/model_parameter_checkpoint.h5',
+                                      verbose=1,
+                                      save_best_only=True,
+                                      period=2,
+                                      save_weights_only=False)
 
-    model_save_path = os.path.join(Path(os.path.abspath(__file__)).parents[2],
-                                   "saved_models",
-                                   training_parameters['architecture'],
-                                   training_timestamp)
+    # We use try to stop the training whenever we want
+    try:
+        history = model.fit_generator(trainDataloader, validation_data=valDataloader,
+                                      epochs=num_of_epochs,
+                                      callbacks=[earlystopper, reduce_lr, hist, modelCheckpoint])
+
+    except KeyboardInterrupt:
+        print('KeyboardInterrupt, do you wanna save the model: yes-(y), no-(n)')
+        save = str(input())
+        if save is 'y':
+            save_history(hist, model_save_path)
+            save_training_parameters(training_parameters, model_save_path)
+            evaluate_training.plot_history(hist, path=model_save_path)
+        else:
+            print(f'Deleting: "{model_save_path}" !')
+            shutil.rmtree(model_save_path)
+        sys.exit(1)
 
-    os.makedirs(model_save_path)
     model.save(model_save_path + '/model_parameter.h5')
+    save_history(history, model_save_path)
+    save_training_parameters(training_parameters, model_save_path)
```

Comment:
Warum speicherst du hier die Training Parameter nochmal extra? So wie ich das [FAQ](https://keras.io/getting-started/faq/#how-can-i-save-a-keras-model) verstehe, wird das durch das `model.save` in Zeile 114 eh schon gemacht. Willst du die nur einzeln abgespeichert haben?

Reply:
Ja genau, weil wir in die `.h5` nicht einfach reinschauen können, aber in eine  `.json` file schon.